File: development/NET/samples/server/Serial_IO/VB/OpcServer.vb
Imports System
Imports System.Collections
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace SerialIO

	Public Class OpcServer

#Region "Constructor"
		'-----------------
		Public Sub New()

		End Sub
		'--
#End Region

#Region "Protected Attributes"
		'--------------------------

		Protected m_shutdownRequest As ShutdownHandler = Nothing
		Protected m_daSimulationElement1 As MyDaAddressSpaceElement = Nothing
		Protected m_daSimulationElement2 As MyDaAddressSpaceElement = Nothing
		Protected m_random As New Random(1)

		Protected m_requestsJanitor As New Object()
		Protected m_requests As New ArrayList()

		'--
#End Region

#Region "Public Properties"
		'-----------------------

		Public Property ServiceName() As String
			Get
				Return Application.Instance.ServiceName
			End Get
			Set(ByVal value As String)
				Application.Instance.ServiceName = value
			End Set
		End Property

		'--
#End Region

#Region "Public Methods"
		'-----------------------

		Public Shared Function Shutdown() As Integer
			Return [Enum].ToObject(GetType(EnumResultCode), EnumResultCode.S_OK)
		End Function

		Public Function Initialize() As System.UInt32
			Try
				Application.Instance.VersionOtb = 447
				Application.Instance.AppType = EnumApplicationType.EXECUTABLE

				Application.Instance.ClsidDa = "{605B206A-6DE0-455b-B4CC-2B3F80324EDC}"
				Application.Instance.ProgIdDa = "Softing.OPC.Toolkit.NET.VB.Sample.SerialIO.DA.1"
				Application.Instance.VerIndProgIdDa = "Softing.OPC.Toolkit.NET.VB.Sample.SerialIO.DA"
				Application.Instance.Description = "Softing OPC Toolkit .NET VB SerialIO Sample Server"
				Application.Instance.MajorVersion = 4
				Application.Instance.MinorVersion = 47
				Application.Instance.PatchVersion = 1
				Application.Instance.BuildNumber = 0
				Application.Instance.VendorInfo = "Softing Industrial Automation GmbH"
				Application.Instance.MinUpdateRateDa = 100
				Application.Instance.ClientCheckPeriod = 30000
				Application.Instance.AddressSpaceDelimiter = "."c
				Application.Instance.PropertyDelimiter = "/"c

				Application.Instance.IpPortHTTP = Convert.ToUInt16(8079)
				Application.Instance.UrlDa = "/OPC/DA"
				AddHandler Application.Instance.ShutdownRequest, AddressOf Shutdown
				Application.Instance.EnableTracing([Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.ALL), [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.ALL), _
				[Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.SERVER), [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.SERVER), "Trace1.txt", System.Convert.ToUInt32(1000000), _
				  System.Convert.ToUInt32(2))
			Catch exc As Exception
				Trace(EnumTraceLevel.ERR, [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.USER1), "OpcServer::Initialize", exc.ToString())

				Return System.Convert.ToUInt32(EnumResultCode.E_FAIL)
			End Try
			Return System.Convert.ToUInt32(EnumResultCode.S_OK)
		End Function

		Public Function Start() As Integer
			Return Application.Instance.Start()
		End Function

		Public Function [Stop]() As Integer
			Return Application.Instance.[Stop]()
		End Function

		Public Function Ready() As Integer
			Return Application.Instance.Ready()
		End Function

		Public Function Terminate() As Integer
			Return Application.Instance.Terminate()
		End Function

		Public Function Prepare(ByVal aMyCreator As MyCreator) As Integer
			Dim result As Integer = Application.Instance.Initialize(aMyCreator)

			Return result
		End Function

		Public Function ProcessCommandLine(ByVal commandLine As String) As Integer
			Return Application.Instance.ProcessCommandLine(commandLine)
		End Function

		Public Function BuildAddressSpace() As System.UInt32
			Try
				Dim creator As MyCreator = CType(Application.Instance.Creator, MyCreator)

				'  DA
				Dim daRoot As DaAddressSpaceRoot = Application.Instance.DaAddressSpaceRoot
				m_daSimulationElement1 = CType(creator.CreateMyDaAddressSpaceElement(), MyDaAddressSpaceElement)
				m_daSimulationElement1.Name = "simulation1"
				m_daSimulationElement1.AccessRights = EnumAccessRights.READWRITEABLE
				m_daSimulationElement1.Datatype = GetType(Integer)
				m_daSimulationElement1.IoMode = EnumIoMode.POLL
				daRoot.AddChild(m_daSimulationElement1)

				m_daSimulationElement2 = CType(creator.CreateMyDaAddressSpaceElement(), MyDaAddressSpaceElement)
				m_daSimulationElement2.Name = "simulation2"
				m_daSimulationElement2.AccessRights = EnumAccessRights.READWRITEABLE
				m_daSimulationElement2.Datatype = GetType(Integer)
				m_daSimulationElement2.IoMode = EnumIoMode.POLL
				daRoot.AddChild(m_daSimulationElement2)
			Catch exc As Exception
				Trace(EnumTraceLevel.ERR, [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.USER1), "OpcServer.BuildAddressSpace", exc.ToString())
				Return Convert.ToUInt32(EnumResultCode.E_FAIL)
			End Try

			Return Convert.ToUInt32(EnumResultCode.S_OK)
		End Function


		Public Sub ChangeSimulationValues()

			If Not m_daSimulationElement1 Is Nothing Then
				m_daSimulationElement1.ValueChanged(New ValueQT(m_random.[Next](5000), [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), DateTime.Now))
			End If

			If Not m_daSimulationElement2 Is Nothing Then
				m_daSimulationElement2.ValueChanged(New ValueQT(m_random.[Next](5000), [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), DateTime.Now))
			End If

		End Sub

		Public Sub AddRequsts(ByVal requests As ArrayList)
			For Each request As MyRequest In requests
				'  set the requests on pending state. 
				'  they will automatically be set to complete state when complete method is called
				request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.PENDING)
			Next
			SyncLock m_requestsJanitor
				m_requests.AddRange(requests)
			End SyncLock
		End Sub


		Public Sub HandleRequests()
			SyncLock m_requestsJanitor
				For Each request As MyRequest In m_requests
					If request.ProgressRequestState(EnumRequestState.PENDING, EnumRequestState.INPROGRESS) Then
						Dim element As MyDaAddressSpaceElement = CType(request.AddressSpaceElement, MyDaAddressSpaceElement)
						If element Is Nothing Then
							request.Result = EnumResultCode.E_FAIL
						Else
							Select Case request.TransactionType
								Case EnumTransactionType.READ
									If request.PropertyId = 0 Then
										Dim cacheValue As ValueQT = Nothing
										request.AddressSpaceElement.GetCacheValue(cacheValue)
										request.Value = cacheValue
										request.Result = [Enum].ToObject(GetType(EnumResultCode), EnumResultCode.S_OK)
									Else
										element.GetPropertyValue(request)
									End If
								Case EnumTransactionType.WRITE
									If request.PropertyId = 0 Then
										Dim newValue As ValueQT = request.Value
										request.Result = [Enum].ToObject(GetType(EnumResultCode), element.ValueChanged(newValue))
									Else
										request.Result = EnumResultCode.E_BADRIGHTS
									End If
								Case Else
							End Select
						End If
					End If
					request.Complete()
				Next
				m_requests.Clear()
			End SyncLock

		End Sub


		Public Sub Trace(ByVal traceLevel As EnumTraceLevel, ByVal traceGroup As EnumTraceGroup, ByVal objectID As String, ByVal message As String)
			Application.Instance.Trace(traceLevel, traceGroup, objectID, message)
		End Sub


		'--
#End Region

	End Class
End Namespace

File: development/NET/samples/server/Serial_IO/CS/OpcServer.cs
using System;
using System.Collections;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace SerialIO
{
	public class OpcServer
	{

		#region Constructor
		//-----------------
		public OpcServer()
		{

		}   //	end constructor
			//--
		#endregion

		#region Protected Attributes
		//--------------------------

		protected ShutdownHandler m_shutdownRequest = null;
		protected MyDaAddressSpaceElement m_daSimulationElement1 = null;
		protected MyDaAddressSpaceElement m_daSimulationElement2 = null;
		protected Random m_random = new Random(1);

		protected object m_requestsJanitor = new object();
		protected ArrayList m_requests = new ArrayList();

		//--
		#endregion

		#region Public Properties
		//-----------------------

		public string ServiceName
		{
			get
			{
				return Application.Instance.ServiceName;
			}
			set
			{
				Application.Instance.ServiceName = value;
			}
		}   //	end SetServiceName

		//--
		#endregion

		#region Public Methods
		//-----------------------

		public static int Shutdown()
		{
			return (int)EnumResultCode.S_OK;
		}   //	end Shutdown

		public int Initialize()
		{
			try
			{
				Application.Instance.VersionOtb = 447;
				Application.Instance.AppType = EnumApplicationType.EXECUTABLE;

				Application.Instance.ClsidDa = "{381057C3-BCEE-4DA5-A05A-66DCE07F3576}";

				Application.Instance.ProgIdDa = "Softing.OPCToolbox.NET.CS.Sample.SerialIO.DA.1";
				Application.Instance.VerIndProgIdDa = "Softing.OPCToolbox.NET.CS.Sample.SerialIO.DA";
				Application.Instance.Description = "Softing OPC Toolkit .NET C# SerialIO Sample Server";
				Application.Instance.MajorVersion = 4;
				Application.Instance.MinorVersion = 47;
				Application.Instance.PatchVersion = 1;
				Application.Instance.BuildNumber = 0;
				Application.Instance.VendorInfo = "Softing Industrial Automation GmbH";
				Application.Instance.MinUpdateRateDa = 100;
				Application.Instance.ClientCheckPeriod = 30000;
				Application.Instance.AddressSpaceDelimiter = '.';
				Application.Instance.PropertyDelimiter = '/';

				Application.Instance.IpPortHTTP = 8079;
				Application.Instance.UrlDa = "/OPC/DA";
				Application.Instance.ShutdownRequest += new Softing.OPCToolbox.Server.ShutdownHandler(Shutdown);
			}
			catch (Exception exc)
			{
				Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER1,
					"OpcServer::Initialize",
					exc.ToString());

				return (int)EnumResultCode.E_FAIL;
			}   //	end try...catch
			return (int)EnumResultCode.S_OK;
		}   //	end Initialize

		public int Start()
		{
			return Application.Instance.Start();
		}   //	end Start

		public int Stop()
		{
			return Application.Instance.Stop();
		}   //	end Stop

		public int Ready()
		{
			return Application.Instance.Ready();
		}   //	end Ready

		public int Terminate()
		{
			return Application.Instance.Terminate();
		}   //	end Terminate

		public int Prepare(MyCreator aMyCreator)
		{
			int result = Application.Instance.Initialize(aMyCreator);

			if (ResultCode.SUCCEEDED(result))
			{
				Application.Instance.EnableTracing(
					EnumTraceGroup.ALL,
					EnumTraceGroup.ALL,
					EnumTraceGroup.SERVER,
					EnumTraceGroup.SERVER,
					"Trace1.txt",
					1000000,
					0);
			}   //	end if

			return result;
		}   //	end Prepare

		public int ProcessCommandLine(string commandLine)
		{
			return Application.Instance.ProcessCommandLine(commandLine);
		}   //	end ProcessCommandLine

		public int BuildAddressSpace()
		{
			try
			{
				MyCreator creator = (MyCreator)Application.Instance.Creator;

				//	DA
				DaAddressSpaceRoot daRoot = Application.Instance.DaAddressSpaceRoot;
				m_daSimulationElement1 = (MyDaAddressSpaceElement)creator.CreateMyDaAddressSpaceElement();
				m_daSimulationElement1.Name = "simulation1";
				m_daSimulationElement1.AccessRights = EnumAccessRights.READWRITEABLE;
				m_daSimulationElement1.Datatype = typeof(System.Int32);
				m_daSimulationElement1.IoMode = EnumIoMode.POLL;
				daRoot.AddChild(m_daSimulationElement1);

				m_daSimulationElement2 = (MyDaAddressSpaceElement)creator.CreateMyDaAddressSpaceElement();
				m_daSimulationElement2.Name = "simulation2";
				m_daSimulationElement2.AccessRights = EnumAccessRights.READWRITEABLE;
				m_daSimulationElement2.Datatype = typeof(System.Int32);
				m_daSimulationElement2.IoMode = EnumIoMode.POLL;
				daRoot.AddChild(m_daSimulationElement2);
			}
			catch (Exception exc)
			{
				Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER1,
					"OpcServer.BuildAddressSpace",
					exc.ToString());
				return (int)EnumResultCode.E_FAIL;
			}   //	end try...catch

			return (int)EnumResultCode.S_OK;
		}   //	end BuildAddressSpace


		public void ChangeSimulationValues()
		{
			if (m_daSimulationElement1 != null)
			{
				m_daSimulationElement1.ValueChanged(new ValueQT(m_random.Next(5000), EnumQuality.GOOD, DateTime.Now));
			}   //	end if

			if (m_daSimulationElement2 != null)
			{
				m_daSimulationElement2.ValueChanged(new ValueQT(m_random.Next(5000), EnumQuality.GOOD, DateTime.Now));
			}   //	end if

		}   //	end ChangeSimulationValues

		public void AddRequsts(ArrayList requests)
		{
			foreach (MyRequest request in requests)
			{
				//	set the requests on pending state. 
				//	they will automatically be set to complete state when complete method is called
				request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.PENDING);
			}
			lock (m_requestsJanitor)
			{
				m_requests.AddRange(requests);
			}
		}


		public void HandleRequests()
		{
			lock (m_requestsJanitor)
			{
				foreach (MyRequest request in m_requests)
				{
					if (request.ProgressRequestState(EnumRequestState.PENDING, EnumRequestState.INPROGRESS))
					{
						MyDaAddressSpaceElement element = request.AddressSpaceElement as MyDaAddressSpaceElement;
						if (element == null)
						{
							request.Result = EnumResultCode.E_FAIL;
						}
						else
						{
							switch (request.TransactionType)
							{
								case EnumTransactionType.READ:
									if (request.PropertyId == 0)
									{
										ValueQT cacheValue = null;
										request.Result = (EnumResultCode)element.GetCacheValue(ref cacheValue);
										request.Value = cacheValue;
									}
									else
									{
										element.GetPropertyValue(request);
									}   //  end if ... else
									break;

								case EnumTransactionType.WRITE:
									if (request.PropertyId == 0)
									{
										ValueQT newValue = request.Value;
										request.Result = (EnumResultCode)element.ValueChanged(newValue);
									}
									else
									{
										request.Result = EnumResultCode.E_BADRIGHTS;
									}   //  end if ... else
									break;

								default:
									break;
							}   //  end switch

						}   //  end if ... else
					}   //  end if ... else
					request.Complete();
				}   //  end forech
				m_requests.Clear();
			}   //  end lock

		}   //  end HandleRequests


		public void Trace(
			EnumTraceLevel traceLevel,
			EnumTraceGroup traceGroup,
			string objectID,
			string message)
		{
			Application.Instance.Trace(
				traceLevel,
				traceGroup,
				objectID,
				message);
		}   //	end Trace


		//--
		#endregion

	}   //	end OpcServer
}   //	end namespace
